File: haskell/glue/src/Glue/Lib/List/Drop.hs
module Glue.Lib.List.Drop where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

drop :: IR Eval
drop = NativeFunc dropImpl

dropImpl :: IR Eval -> Eval (IR Eval)
dropImpl countIR = pure $ NativeFunc (dropFrom countIR)

dropFrom :: IR Eval -> IR Eval -> Eval (IR Eval)
dropFrom count list = case (count, list) of
    (Integer n, List xs) -> do
        if n < 0
            then throwError $ wrongArgumentType ["non-negative integer"]
            else pure $ List (Prelude.drop (fromIntegral n) xs)
    _ -> throwError $ wrongArgumentType ["number", "list"]

File: dart/glue/lib/src/lib/list/drop.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Drop function - removes first N elements from a list
/// Mirrors Haskell Glue.Lib.List.Drop.drop exactly
Ir drop = IrNativeFunc(dropImpl);

/// Drop function implementation
/// Mirrors Haskell Glue.Lib.List.Drop.dropImpl exactly
Eval<Ir> dropImpl(Ir countIr) {
  return Eval.pure(IrNativeFunc(dropFrom(countIr)));
}

/// Helper function for list argument
/// Mirrors Haskell Glue.Lib.List.Drop.dropFrom exactly
Eval<Ir> Function(Ir) dropFrom(Ir countIr) {
  return (Ir listIr) => switch ((countIr, listIr)) {
    (IrInteger(value: final n), IrList(elements: final elements)) =>
      n < 0
          ? throwError(wrongArgumentType(['non-negative integer']))
          : Eval.pure(
              IrList(n >= elements.length ? <Ir>[] : elements.skip(n).toList()),
            ),
    _ => throwError(wrongArgumentType(['number', 'list'])),
  };
}
